libs/providers/langchain-mcp-adapters/src/interceptor.ts

```diff
@@ -0,0 +1,168 @@
+
+export { RunnableConfig };
+
+const toolCallRequestSchema = z.object({
+  serverName: z.string(),
+  name: z.string(),
+  args: z.unknown(),
+});
+export type ToolCallRequest = z.output<typeof toolCallRequestSchema>;
+
+const toolResultSchema = z.tuple([
+  z.custom<string | (ContentBlock | ContentBlock.Data.DataContentBlock)[]>(),
+  z.array(
+    z.union([
+      EmbeddedResourceSchema,
+      z.custom<ContentBlock.Data.DataContentBlock>(),
+    ])
+  ),
+]);
+export type ToolResult = z.output<typeof toolResultSchema>;
+
+const toolCallResultSchema = z.object({
+  ...toolCallRequestSchema.shape,
+  result: toolResultSchema,
+});
+export type ToolCallResult = z.output<typeof toolCallResultSchema>;
+
+const toolCallModificationSchema = z
+  .object({
+    header: z.record(z.string()),
+    args: z.unknown(),
+  })
+  .partial();
+export type ToolCallModification = z.output<typeof toolCallModificationSchema>;
+
+export const interceptorSchema = z.object({
+  /**
+   * Called before a tool call is made.
+   * Allows you to modify the tool call arguments or return a different tool call.
+   *
+   * @param toolCallRequest - The tool call request
+   * @param toolCallRequest.name - The tool name
+   * @param toolCallRequest.args - The tool call arguments
+   * @param toolCallRequest.serverName - The server name
+   * @param config - The runnable config
+   * @returns The tool call modification
+   *
+   * @example
+   * ```ts
+   * const interceptor = {
+   *   beforeToolCall: (toolCallRequest) => {
+   *     return {
+   *       args: {
+   *         ...toolCallRequest.args,
+   *         custom: "Custom Value"
+   *       },
+   *       header: { "X-Custom-Header": "Custom Value" }
+   *     };
+   *   },
+   * };
+   * ```
+   */
+  beforeToolCall: z
+    .function()
+    .args(toolCallRequestSchema, z.custom<RunnableConfig>())
+    .returns(
+      z.union([
+        z.promise(toolCallModificationSchema),
+        toolCallModificationSchema,
+        z.void(),
+        z.promise(z.void()),
+      ])
+    )
+    .optional(),
+
+  /**
+   * Called after a tool call is made.
+   * Allows you to modify the tool call result or return a different tool call result.
+   *
+   * @param toolCallResult - The tool call result
+   * @param toolCallResult.args - The tool call arguments
+   * @param toolCallResult.serverName - The server name
+   * @param toolCallResult.name - The tool name
+   * @param toolCallResult.result - The tool call result
+   * @param config - The runnable config
+   * @returns The tool call modification
+   * @example
+   * ```ts
+   * const interceptor = {
+   *   afterToolCall: (toolCallResult) => {
+   *     if (toolCallResult.name === "calculator") {
+   *       return ["Custom Value", []];
+   *     }
+   *     return toolCallResult.result;
+   *   },
+   * };
+   * ```
+   */
+  afterToolCall: z
+    .function()
+    .args(toolCallResultSchema, z.custom<RunnableConfig>())
+    .returns(
```

Comment:
langchain tools can support:

1. Command() when running within langgraph
2. 2-tuple of content, artifact
3. ToolMessage return (in some cases)

Have you thought about how we could support these? The main one I think is `Command()` in case users want to update agent state from the tool call (it's a fairly common pattern)


The 2-tuple of content, artifact might be a good way to allow users to customize handling structured content?

Reply:
> langchain tools can support:

Should we also support just a `string`?

Reply:
yeah definitely :)

Reply:
added string support